### PR TITLE
Migrate to entry_point macro

### LINKED ISSUE #230: Use #[entry_point] macro in contracts
This is now the most flexible way to do it and has been demoed in #216

Let's adopt it for the contracts (note we need to add some cfg statement so we only compile if wasm target AND not library feature) This is demoed in the linked PR

### PR DESCRIPTION
**Pull request by maurolacy**
Closes #230.

### CODE REVIEW

File: contracts/cw1-subkeys/src/contract.rs [hunk 1]
```diff
@@ -30,6 +30,7 @@ use cw_storage_plus::Bound;
 const CONTRACT_NAME: &str = "crates.io:cw1-subkeys";
 const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 
+#[entry_point]
```
Comment (ethanfrey) [inline contracts/cw1-subkeys/src/contract.rs:33]: Can you use `#[cfg_attr(not(feature = "library"), entry_point)]` as you did for whitelist?

File: contracts/cw1-whitelist/src/contract.rs [hunk 2]
```diff
@@ -16,6 +19,7 @@ use crate::state::{AdminList, ADMIN_LIST};
 const CONTRACT_NAME: &str = "crates.io:cw1-whitelist";
 const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 
+#[cfg_attr(not(feature = "library"), entry_point)]
```
Comment (ethanfrey) [inline contracts/cw1-whitelist/src/contract.rs:22]: This is proper, and should be the default for all, whether or not se currently import them as libraries, they should be importable as libraries.

File: contracts/cw1-whitelist/src/contract.rs [hunk 3]
```diff
@@ -1,10 +1,13 @@
 use schemars::JsonSchema;
 use std::fmt;
 
+#[cfg(not(feature = "library"))]
```
Comment (ethanfrey) [inline contracts/cw1-whitelist/src/contract.rs:4]: Very good point here, I think I messed up in other contracts, making this a default import.

File: contracts/cw20-bonding/src/contract.rs [hunk 4]
```diff
@@ -23,6 +23,7 @@ use cw0::{must_pay, nonpayable};
 const CONTRACT_NAME: &str = "crates.io:cw20-bonding";
 const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
 
+#[entry_point]
```
Comment (ethanfrey) [inline contracts/cw20-bonding/src/contract.rs:26]: As above, please gate all of these `entry_point`s in all contracts with library feature

**Review by ethanfrey — CHANGES_REQUESTED** [review]
Please don't use the plain `#[entry_point]` decorator, but rather the library feature flag version. I actually asked Simon to make that the default functionality (disable on library flag), but he thought that was putting too much app-specific logic in cosmwasm base. 

If you find the current status too annoying, please look at the original cosmwasm PR; https://github.com/CosmWasm/cosmwasm/pull/701#issuecomment-757221171 and re-open this issue. I just accepted using this larger macro heading everywhere.

File: contracts/cw1-subkeys/src/contract.rs (hunk 1 above)
Comment (maurolacy) [reply contracts/cw1-subkeys/src/contract.rs:33]: Sure. I did that only when currently necessary. I agree that it's more robust / forward-looking to do it by default.

File: contracts/cw1-whitelist/src/contract.rs (hunk 3 above)
Comment (maurolacy) [reply contracts/cw1-whitelist/src/contract.rs:4]: Yes, this triggers an `unused_imports` warning if not present, when importing as library.

**Comment by maurolacy** [comment]
> If you find the current status too annoying, please look at the original cosmwasm PR; [CosmWasm/cosmwasm#701 (comment)](https://github.com/CosmWasm/cosmwasm/pull/701#issuecomment-757221171) and re-open this issue. I just accepted using this larger macro heading everywhere.

It's OK with me. Maybe the macro can be written in a way that has this check inside? (just curious)

**Comment by maurolacy** [comment]
Will create a new branch to migrate cw-plus to Rust 0.51.0, as there are a number of linting issues to deal with.

**Comment by ethanfrey** [comment]
> It's OK with me. Maybe the macro can be written in a way that has this check inside? (just curious)

Of course there is, the macro could also emit nothing if the "library" feature flag was set. Simon was worried that we would then be adding "magic" feature flag support into the macros, and said to do that in this repo. I see his point (make it explicit) and your point (make it easy)

**Comment by ethanfrey** [comment]
> Will create a new branch to migrate cw-plus to Rust 0.51.0, as there are a number of linting issues to deal with.

Sounds good. So, you will do that first and then rebase this on it? I will approve it, so as soon as CI passes, I am happy.

**Review by ethanfrey — APPROVED** [review]
Looks good. As soon as the CI passes (I think the build script just needs to be updated to Rust 1.51)

**Comment by maurolacy** [comment]
This was caused by thec `check_contract` dep being built from main istead of from a tag.

Changed it so we can merge this independently of the Rust 1.51.0 migration (than @ethanfrey).

File: .circleci/config.yml [hunk 5]
```diff
@@ -834,7 +834,7 @@ jobs:
           # Uses --debug for compilation speed
           # FIXME: Change when `check_contract` (part of `cosmwasm-0.14.0`) is published
           #command: cargo install --debug --features iterator --example check_contract -- cosmwasm-vm
-          command: cargo install --debug --features iterator --git https://github.com/CosmWasm/cosmwasm --branch=main --example check_contract -- cosmwasm-vm
+          command: cargo install --debug --features iterator --git https://github.com/CosmWasm/cosmwasm --tag=v0.14.0-beta1 --example check_contract -- cosmwasm-vm
```
Comment (ethanfrey) [inline .circleci/config.yml:837]: :rocket:

**Pull request merged**